# DEPENDENCIES AND SETUP

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np
import os
import csv
from IPython.display import HTML

## LOAD CSV and DISPLAY


In [2]:
# File to Load - CHANGE FILE PATH and INCLUDE os.path.join 
# file_to_load = "Resources/purchase_data.csv"
csvpath = os.path.join("purchase_data.csv")
print(csvpath)


# Read Purchasing File and store into Pandas data frame
purchase_df = pd.read_csv(csvpath)
purchase_df.head()

purchase_data.csv


Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
# using describe as a counter check during calculations
    #using min and max in later calculations to evaluate age bins
purchase_df.describe()

Purchase ID         Age     Item ID       Price
count   780.000000  780.000000  780.000000  780.000000
mean    389.500000   22.714103   92.114103    3.050987
std     225.310896    6.659444   52.775943    1.169549
min       0.000000    7.000000    0.000000    1.000000
25%     194.750000   20.000000   48.000000    1.980000
50%     389.500000   22.000000   93.000000    3.150000
75%     584.250000   25.000000  139.000000    4.080000
max     779.000000   45.000000  183.000000    4.990000

## DISPLAY TOTAL PLAYER COUNT

In [4]:
# Display the total number of players (using unique as this is only total that has value versus total rows of players)

player_count = len(purchase_df["SN"].unique())
print(player_count)

576


## PURCHASING ANALYSIS

In [5]:
# Purchasing Analysis (Total)
# Run basic calculations to obtain number of unique items, average price, etc.

#unique item count
item_count = len(purchase_df["Item ID"].unique())

#Average price
    #price_average = (purchase_df["Price"].count()/)
average_price = [np.round(purchase_df["Price"].mean(), 2)]

#Total number of purchases
purchase_count = len(purchase_df["Purchase ID"].unique())

#Total revenue
total_revenue = (purchase_df["Price"].sum())

# Create a summary data frame to hold the results
summary_table = pd.DataFrame({"Total Unique Players": player_count,
                              "Total Unique Items": [item_count],
                              "Average Price": [average_price],
                              "Number of Purchases": [purchase_count],
                              "Total Revenue":[total_revenue]})   

summary_table.style.format({"Average Price": "${:.2f}", "Total Revenue": "${:.2f}"})  

# Optional: give the displayed data cleaner formatting, I FOUND on Pandas.pydata.org, 
# but cannot get it merged into other formatting I like it and would like to use it but want the Price
# and revenue columns formatted too
def hover(hover_color="#ffff99"):
    return dict(selector="tr:hover",
                props=[("background-color", "%s" % hover_color)])

styles = [
    hover(),
    dict(selector="th", props=[("font-size", "150%"),
                               ("text-align", "center")]),
    dict(selector="caption", props=[("caption-side", "bottom")])
]
html = (summary_table.style.set_table_styles(styles)
          .set_caption("Hover to highlight."))
html

# Display the summary data frame


In [6]:
#Gener Demographics
#get a unique count of players by gender to validate the specific counts
gender_uniquecount = purchase_df.groupby(["Gender"])["SN"].nunique()
gender_uniquecount

Gender
Female                    81
Male                     484
Other / Non-Disclosed     11
Name: SN, dtype: int64

## GENDER DEMOGRAPHICS

In [7]:
#Gender Demographics

#Percentage and Count of Female Player
female_gender = purchase_df[purchase_df["Gender"] == "Female"]["SN"].nunique()

#Percentage and Count of Male Players
male_gender = purchase_df[purchase_df["Gender"] == "Male"]["SN"].nunique()

#Percentage and Count of Other / Non-Disclosed
other_gender = player_count - male_gender - female_gender

female_percent = ((female_gender / player_count)*100)
male_percent = ((male_gender / player_count)*100)
other_percent = ((other_gender / player_count)*100)

#table of gender data

gender_df = pd.DataFrame({"Gender": ["Female", "Male", "Other / Non-Disclosed"], "Percentage of Players":
                                      [female_percent, male_percent, other_percent],
                                      "Total Count": [female_gender, male_gender, other_gender]}, columns = 
                                       ["Gender", "Percentage of Players", "Total Count"])
                                        
gender_df_final = gender_df.set_index("Gender")
gender_df_final.style.format({"Percentage of Players": "{:.2f}%"})     

## PURCHASING ANALYSIS BY GENDER

In [8]:
# Purchasing Analysis (Gender)

#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender

#Counts by gender
female_purchase = purchase_df[purchase_df["Gender"] == "Female"]["Price"].count()
male_purchase = purchase_df[purchase_df["Gender"] == "Male"]["Price"].count()
other_purchase = purchase_df[purchase_df["Gender"] == "Other / Non-Disclosed"]["Price"].count()

#Averages by gender
female_pa = purchase_df[purchase_df["Gender"] == "Female"]['Price'].mean()
male_pa = purchase_df[purchase_df["Gender"] == "Male"]['Price'].mean()
other_pa = purchase_df[purchase_df["Gender"] == "Other / Non-Disclosed"]['Price'].mean()

#Sum by gender
female_sum = purchase_df[purchase_df["Gender"] == "Female"]['Price'].sum()
male_sum = purchase_df[purchase_df["Gender"] == "Male"]['Price'].sum()
other_sum = purchase_df[purchase_df["Gender"] == "Other / Non-Disclosed"]['Price'].sum()

#Average Total Purchase per person by Gender
female_total = female_sum/female_purchase
male_total = male_sum/male_purchase
other_total = other_sum/other_purchase

#Create a summary data frame to hold the results

#Optional: give the displayed data cleaner formatting (figure out how to do the special formatting to ALL!)


gender_purchase_df = pd.DataFrame({"Gender": ["Female", "Male", "Other / Non-Disclosed"], 
                                   "Purchase Count": [female_purchase, male_purchase, other_purchase],
                                    "Average Purchase Price": [female_pa, male_pa, other_pa], 
                                    "Total Purchase Value": [female_sum, male_sum, other_sum],
                                    "Average Total Purchase": [female_total, male_total, other_total]}, 
                                     columns = ["Gender", "Purchase Count", "Average Purchase Price", 
                                                "Total Purchase Value", "Average Total Purchase"])
                                        
gender_purchase_final = gender_purchase_df.set_index("Gender")
gender_purchase_final.style.format({"Average Purchase Price": "${:.2f}", 
                                    "Total Purchase Value": "${:.2f}", "Average Total Purchase": "${:.2f}"})

#Display the summary data frame

## AGE DEMOGRAPHICS

In [9]:
#Age Demographics

#Establish bins for ages  MIN was 7 MAX was 45
#Categorize the existing players using the age bins. ?? Hint: use pd.cut()??
# CREATE ACTUAL BINS  (I realized this too late and had to fix 6-4-19.  The way I did it worked, 
# however, I couldn't then get the next group to work.)  I commented out, what I started with. 
# my apologies for the messy, but I wanted to learn from this error. 

#Calculate the numbers and percentages by age group

#underten = purchase_df[purchase_df["Age"] <10]
#lowteens = purchase_df[(purchase_df["Age"] >=10) & (purchase_df["Age"] <=14)]
#hiteens = purchase_df[(purchase_df["Age"] >=15) & (purchase_df["Age"] <=19)]
#lotwenty = purchase_df[(purchase_df["Age"] >=20) & (purchase_df["Age"] <=24)]
#hitwenty = purchase_df[(purchase_df["Age"] >=25) & (purchase_df["Age"] <=29)]
#lothirty = purchase_df[(purchase_df["Age"] >=30) & (purchase_df["Age"] <=34)]
#hithirty = purchase_df[(purchase_df["Age"] >=35) & (purchase_df["Age"] <=39)]
#loforty = purchase_df[(purchase_df["Age"] >=40) & (purchase_df["Age"] <=44)]
#hiforty = purchase_df[(purchase_df["Age"] >=45) & (purchase_df["Age"] <=49)]


#Create a summary data frame to hold the results
#age_df = pd.DataFrame({"Age": ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40-44", "45-49"],
 #                       "Percentage of Players": [(underten["SN"].nunique()/player_count)*100, 
  #                                                (lowteens["SN"].nunique()/player_count)*100, 
   #                                               (hiteens["SN"].nunique()/player_count)*100, 
    #                                              (lotwenty["SN"].nunique()/player_count)*100, 
     #                                             (hitwenty["SN"].nunique()/player_count)*100,
      #                                            (lothirty["SN"].nunique()/player_count)*100,
       #                                           (hithirty["SN"].nunique()/player_count)*100, 
        #                                          (loforty["SN"].nunique()/player_count)*100, 
         #                                         (hiforty["SN"].nunique()/player_count)*100],
          #              "Total Count": [underten["SN"].nunique(), lowteens["SN"].nunique(), 
           #                             hiteens["SN"].nunique(), lotwenty["SN"].nunique(), 
            #                            hitwenty["SN"].nunique(), lothirty["SN"].nunique(), 
             #                           hithirty["SN"].nunique(), loforty["SN"].nunique(), 
              #                          hiforty["SN"].nunique()]
               #        })

#Optional: round the percentage column to two decimal points

#age_final = age_df.set_index("Age")
#age_final.style.format({"Percentage of Players": "{:.2f}%"})  
  
#Display Age Demographics Table
age_bins = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Segment and sort age values into bins established above
purchase_df["Age Group"] = pd.cut(purchase_df["Age"],age_bins, labels=group_names)
purchase_df

# Create new data frame with the added "Age Group" and group it
age_grouped = purchase_df.groupby("Age Group")

# Count total players by age category
total_count_age = age_grouped["SN"].nunique()

# Calculate percentages by age category 
percentage_by_age = (total_count_age/player_count) * 100

# Create data frame with obtained values
age_demographics = pd.DataFrame({"Percentage of Players": percentage_by_age, "Total Count": total_count_age})

# Format the data frame with no index name in the corner
age_demographics.index.name = None

# Format percentage with two decimal places 
age_demographics.style.format({"Percentage of Players":"{:,.2f}"})

## PURCHASING ANALYSIS BY AGE

In [10]:
# Purchasing Analysis (Age)

# Bin the purchase_data data frame by age  CREATING BINS -  (This time I read the instructions better)
# Count purchases by age group
purchase_count_age = age_grouped["Purchase ID"].count()

# Obtain average purchase price by age group 
average_purchase_price_age = age_grouped["Price"].mean()

# Calculate total purchase value by age group 
total_purchase_value = age_grouped["Price"].sum()

# Calculate the average purchase per person in the age group 
average_purchase_per_person_age = total_purchase_value/total_count_age

# Create data frame with obtained values
age_demographics = pd.DataFrame({"Purchase Count": purchase_count_age,
                                 "Average Purchase Price": average_purchase_price_age,
                                 "Total Purchase Value":total_purchase_value,
                                 "Average Purchase Total per Person": average_purchase_per_person_age})

# Format the data frame with no index name in the corner
age_demographics.index.name = None

# Format with currency style
age_demographics.style.format({"Average Purchase Price":"${:,.2f}",
                               "Total Purchase Value":"${:,.2f}",
                               "Average Purchase Total per Person":"${:,.2f}"})


## Top Spenders

In [11]:
#Top Spenders

#Run basic calculations to obtain the results in the table below, 
#Purchase Count	Average

# Calculate 'Total Purchase Count' for each SN:  count the SN and price values from the DF
count_per_SN = purchase_df.groupby('SN')['Price'].count()
count_per_SN_df = pd.DataFrame(count_per_SN)

# Calculate 'Total Purchase Price' for each SN: find the unique of the SN values within the DF and sum the prices
unique_SN_df = purchase_df.groupby('SN')['Price'].sum()

#Purchase Price	Total Purchase Value	Avg 
#Total Purchase per Person
# Calculate 'Average Purchase Price' Use mean to find the average of the SN spend within the DF
avg_price_per_SN = purchase_df.groupby('SN')['Price'].mean()

#Create a summary data frame to hold the results

# Add 'Average Purchase Price' to table  defining the column Average purchase
count_per_SN_df['Average Purchase Price'] = avg_price_per_SN

# Add 'Total Purchase Value' to table defining the column total purchase
count_per_SN_df['Total Purchase Value'] = unique_SN_df
count_per_SN_df

# Rename 'Price' to 'Purchase Count'  
updated_count_per_SN_df = count_per_SN_df.rename(columns={
    'Price': 'Purchase Count',
})

#Sort the total purchase value column in descending order

# Sort 'Price' from highest to lowest
sorted_SN_df = updated_count_per_SN_df.sort_values(by=['Total Purchase Value'], ascending=False)
sorted_SN_df

#Optional: give the displayed data cleaner formatting

#Display a preview of the summary data frame

# Show top 5 spenders only
sorted_SN_df.head(5)





Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                3.792000                 18.96
Idastidru52               4                3.862500                 15.45
Chamjask73                3                4.610000                 13.83
Iral74                    4                3.405000                 13.62
Iskadarya95               3                4.366667                 13.10

## Most Popular Items

In [12]:
#Most Popular Items  I'm concerned in the most popular and in the most profitable, I do not get the same values as the sample. 

#Retrieve the Item ID, Item Name, and Item Price columns

#Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
popular_total = purchase_df.groupby("Item Name")["Price"].sum()
popular_count = purchase_df.groupby("Item Name")["Price"].count()
popular_price = purchase_df.groupby("Item Name")["Price"].mean()

#Create a summary data frame to hold the results
popular_df = pd.DataFrame({"Purchase Count": popular_count, "Average Item Price": 
                           popular_price, "Total Purchase Value": popular_total})

#Optional: give the displayed data cleaner formatting

# Sort the purchase count column in descending order 
popular_df.sort_values("Purchase Count", ascending = False, inplace=True)

# Format columns
popular_df["Average Item Price"] = popular_df["Average Item Price"].map("${:.2f}".format)
popular_df["Total Purchase Value"] = popular_df["Total Purchase Value"].map("${:.2f}".format)

# Show Top 5 only
popular_df.head()


Purchase Count  \
Item Name                                                      
Final Critic                                              13   
Oathbreaker, Last Hope of the Breaking Storm              12   
Persuasion                                                 9   
Nirvana                                                    9   
Extraction, Quickblade Of Trembling Hands                  9   

                                             Average Item Price  \
Item Name                                                         
Final Critic                                              $4.61   
Oathbreaker, Last Hope of the Breaking Storm              $4.23   
Persuasion                                                $3.22   
Nirvana                                                   $4.90   
Extraction, Quickblade Of Trembling Hands                 $3.53   

                                             Total Purchase Value  
Item Name                                                          
Final Critic                                               $59.99  
Oathbreaker, Last Hope of the Breaking Storm               $50.76  
Persuasion                                                 $28.99  
Nirvana                                                    $44.10  
Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

In [13]:
#Most Profitable Items

#Sort the above table by total purchase value in descending order
profitable_df = pd.DataFrame({"Purchase Count": popular_count, "Average Item Price": popular_price,
                              "Total Purchase Value": popular_total})

# Sort by Total Purchase Value
profitable_df.sort_values("Total Purchase Value", ascending = False, inplace=True)
# Format columns
profitable_df["Average Item Price"] = profitable_df["Average Item Price"].map("${:.2f}".format)
profitable_df["Total Purchase Value"] = profitable_df["Total Purchase Value"].map("${:.2f}".format)

#Optional: give the displayed data cleaner formatting

#Display a preview of the data frame

# Show Top 5 only
profitable_df.head()

Purchase Count  \
Item Name                                                      
Final Critic                                              13   
Oathbreaker, Last Hope of the Breaking Storm              12   
Nirvana                                                    9   
Fiery Glass Crusader                                       9   
Singed Scalpel                                             8   

                                             Average Item Price  \
Item Name                                                         
Final Critic                                              $4.61   
Oathbreaker, Last Hope of the Breaking Storm              $4.23   
Nirvana                                                   $4.90   
Fiery Glass Crusader                                      $4.58   
Singed Scalpel                                            $4.35   

                                             Total Purchase Value  
Item Name                                                          
Final Critic                                               $59.99  
Oathbreaker, Last Hope of the Breaking Storm               $50.76  
Nirvana                                                    $44.10  
Fiery Glass Crusader                                       $41.22  
Singed Scalpel                                             $34.80

# Final Summary

Summary of the data for Pymoli.
A simple describe of the data shows that the minimum age of seven and a maximum age of 45, as well as the minimum price of $1.00 and a maximum price of $4.99.  The total number of unique players, 576 and a total of 780 purchases, is small total for the game considering The Sims 3 had a total of 7.95 million units sold as of January 2018.  It is valuable to note that of the 576 players 484 were male and only 81 were female, an 84% male to 14% female ratio.  Current gamer statistics (2018) showed the percentages of male to female players much closer, with 45% of players being female and 55% male, as a part of the whole. (https://www.statista.com/statistics/232383/gender-split-of-us-computer-and-video-gamers/)
The total purchase value of the male versus female for Pymoli showed that the male purchaser may have added more value, the female purchaser was willing to spend slighly more, per purchase. In other gender related world news, it demonstrates that female players are more likely to play when given content driven by their motivating factors, which vary slightly from male. (https://www.wepc.com/news/video-game-statistics/).
Age related information clearly shows a preference in the 20 - 24 age grouping while the average age of the average gamer in 2018 is actually 34 years of age.  This game clearly shows it is preferred by a younger male audience than that of the world average gamer. 
